File: api/images/generate.ts
import type { VercelRequest, VercelResponse } from "@vercel/node";

const GEMINIGEN_API_URL = "https://api.geminigen.ai/uapi/v1/generate_image";

export default async function handler(req: VercelRequest, res: VercelResponse) {
  res.setHeader("Access-Control-Allow-Origin", "*");
  res.setHeader("Access-Control-Allow-Methods", "POST, OPTIONS");
  res.setHeader("Access-Control-Allow-Headers", "Content-Type");

  if (req.method === "OPTIONS") {
    return res.status(200).end();
  }
  if (req.method !== "POST") {
    return res.status(405).json({ error: "Method not allowed" });
  }

  const apiKey = process.env.GEMINIGEN_API_KEY;
  if (!apiKey) {
    return res.status(500).json({ error: "GEMINIGEN_API_KEY not configured" });
  }

  const { prompt, model = "imagen-flash", aspect_ratio = "1:1", style } = req.body;
  if (!prompt) {
    return res.status(400).json({ error: "Missing prompt" });
  }

  try {
    const formBody = new URLSearchParams();
    formBody.append("prompt", prompt);
    formBody.append("model", model);
    formBody.append("aspect_ratio", aspect_ratio);
    if (style && style !== "None" && style !== "none") {
      formBody.append("style", style);
    }

    const resp = await fetch(GEMINIGEN_API_URL, {
      method: "POST",
      headers: {
        "x-api-key": apiKey,
        "Accept": "application/json",
      },
      body: formBody,
    });

    if (!resp.ok) {
      if (resp.status >= 400 && resp.status < 500) {
        try {
          const errData = await resp.json();
          const errMsg = errData?.detail?.error_message || `Request failed (${resp.status})`;
          return res.status(resp.status).json({ error: errMsg });
        } catch {
          return res.status(resp.status).json({ error: `Client error: ${resp.status}` });
        }
      }
      const errText = await resp.text().catch(() => "");
      return res.status(resp.status).json({
        error: `GeminiGen API error (${resp.status})`,
        details: errText.slice(0, 500),
      });
    }

    const result = await resp.json();

    // Transform GeminiGen response to match frontend expected format
    // GeminiGen returns { base64_images: "..." }
    if (result.base64_images) {
      return res.status(200).json({
        data: [{ url: `data:image/png;base64,${result.base64_images}` }],
      });
    }

    // Fallback: return raw response
    return res.status(200).json(result);
  } catch (err) {
    return res
      .status(500)
      .json({ error: "Failed to call GeminiGen API", details: String(err) });
  }
}

File: api/images/generate.py
from http.server import BaseHTTPRequestHandler
import json
import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..'))
from grok_lib.client import Grok


class handler(BaseHTTPRequestHandler):

    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

    def do_POST(self):
        try:
            content_length = int(self.headers.get('Content-Length', 0))
            body = json.loads(self.rfile.read(content_length)) if content_length else {}

            prompt = body.get('prompt')
            if not prompt:
                self._send_json(400, {"error": "Missing prompt"})
                return

            model = body.get('model', 'grok-3')
            image_count = body.get('imageGenerationCount', 2)

            grok = Grok(model=model)
            result = grok.start_convo(
                message=f"Generate an image: {prompt}",
                enableImageGeneration=True,
                imageGenerationCount=image_count,
                forceConcise=True,
                disableTextFollowUps=True,
            )

            if result.get('error'):
                self._send_json(500, {"error": result['error']})
                return

            image_urls = result.get("images") or []
            if image_urls:
                self._send_json(200, {
                    "data": [{"url": url} for url in image_urls],
                })
            else:
                self._send_json(200, {
                    "data": [],
                    "message": result.get("response") or "No images were generated. Try a different prompt.",
                })

        except Exception as e:
            self._send_json(500, {"error": "Failed to generate image", "details": str(e)})

    def _send_json(self, status_code, data):
        self.send_response(status_code)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        self.wfile.write(json.dumps(data).encode())
